From 0edaea8f91bf08025651eeff32f53b9335003842 Mon Sep 17 00:00:00 2001
From: "Richard W.M. Jones" <rjones@redhat.com>
Date: Tue, 8 Oct 2024 14:53:38 +0100
Subject: [PATCH] website: Add links for 1.54 and 1.55 download locations

---
 website/index.html.in | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/website/index.html.in b/website/index.html.in
index a61fca92..2a0d3221 100644
--- a/website/index.html.in
+++ b/website/index.html.in
@@ -186,9 +186,10 @@ git clone <a href="https://github.com/libguestfs/libguestfs">https://github.com/
 LATEST-URL: http://libguestfs.org/download/1.53-development/libguestfs-@PACKAGE_VERSION@.tar.gz
 LATEST-VERSION: @PACKAGE_VERSION@
   -->
-  <a href="download/1.53-development/">Latest development version: <strong>@PACKAGE_VERSION@</strong></a> (released <strong>@RELEASE_DATE@</strong>).<br/>
+  <a href="download/1.55-development/">Latest development version: <strong>@PACKAGE_VERSION@</strong></a> (released <strong>@RELEASE_DATE@</strong>).<br/>
   Stable branch:
   <strong>
+  <a href="download/1.54-stable/">1.54.x</a>,
   <a href="download/1.52-stable/">1.52.x</a>,
   <a href="download/1.50-stable/">1.50.x</a>,
   <a href="download/1.48-stable/">1.48.x</a>
